kubernetes-sigs/kustomize
# Add `--helm-debug` Flag to Kustomize for Enhanced Helm Debugging

### **Summary of the Enhancement**
This PR adds a new flag, `--helm-debug`, to the Kustomize CLI. When this flag is provided, the `--debug` flag is passed to the Helm executable. This enables more detailed logging and error information from Helm, which is helpful when diagnosing issues with Helm charts during the Kustomize build process.

### **Why This is Useful**
In cases where Helm chart templates fail to render correctly, the default error messages can be opaque and difficult to diagnose. By enabling Helm’s `--debug` flag via `--helm-debug`, users can see detailed logs from Helm, including the specific files and lines where errors occur, as well as the internal processes Helm is performing. This can save significant time in troubleshooting and resolving issues.

### **Example Usage**
Without `--helm-debug`:
```sh
kustomize build . --load-restrictor=LoadRestrictionsNone --enable-helm
```

Output:
```
Error: Error: YAML parse error on s3/templates/bucket.yaml: error converting YAML to JSON: yaml: line 9: found character that cannot start any token

Use --debug flag to render out invalid YAML
: unable to run: 'helm template s3...'
```

With `--helm-debug`:
```sh
kustomize build . --load-restrictor=LoadRestrictionsNone --enable-helm --helm-debug
```

Output:
```
Error: install.go:218: [debug] Original chart version: ""
install.go:235: [debug] CHART PATH: /Users/isarn/git/IAC/crossplane/devops/live/devops/prod/ca-central-1/s3/iaas-scripts/charts/s3-0.1.1/s3

Error: YAML parse error on s3/templates/bucket.yaml: error converting YAML to JSON: yaml: line 9: found character that cannot start any token
helm.go:84: [debug] error converting YAML to JSON: yaml: line 9: found character that cannot start any token
YAML parse error on s3/templates/bucket.yaml
helm.sh/helm/v3/pkg/releaseutil.(*manifestFile).sort
        helm.sh/helm/v3/pkg/releaseutil/manifest_sorter.go:146
helm.sh/helm/v3/pkg/releaseutil.SortManifests
        helm.sh/helm/v3/pkg/releaseutil/manifest_sorter.go:106
helm.sh/helm/v3/pkg/action.(*Configuration).renderResources
        helm.sh/helm/v3/pkg/action/action.go:168
helm.sh/helm/v3/pkg/action.(*Install).RunWithContext
        helm.sh/helm/v3/pkg/action/install.go:304
main.runInstall
        helm.sh/helm/v3/cmd/helm/install.go:310
main.newTemplateCmd.func2
        helm.sh/helm/v3/cmd/helm/template.go:95
github.com/spf13/cobra.(*Command).execute
        github.com/spf13/cobra@v1.8.0/command.go:983
github.com/spf13/cobra.(*Command).ExecuteC
        github.com/spf13/cobra@v1.8.0/command.go:1115
github.com/spf13/cobra.(*Command).Execute
        github.com/spf13/cobra@v1.8.0/command.go:1039
main.main
        helm.sh/helm/v3/cmd/helm/helm.go:83
runtime.main
        runtime/proc.go:271
runtime.goexit
        runtime/asm_amd64.s:1695
---
# Source: s3/templates/bucket.yaml
apiVersion: s3.aws.upbound.io/v1beta1
kind: Bucket
metadata:
  labels:
    project: 
    env: 
    region: us-west-2
    crossplane-name: my-bucket
  name: %!!(MISSING)s(<nil>)-%!!(MISSING)s(<nil>)-usw2-my-bucket
spec:
  forProvider:
    tags:
      Name: %!!(MISSING)s(<nil>)-%!!(MISSING)s(<nil>)-usw2-my-bucket
  providerConfigRef:
    name:%!!(MISSING)s(<nil>)-%!!(MISSING)s(<nil>)
: unable to run: 'helm template s3 ...'
...
```

### **How to Test**
1. Add the `--helm-debug` flag to your Kustomize command as shown above.
2. Run the command against a Helm chart with known issues.
3. Verify that the detailed debug output is displayed, helping to identify the root cause of the error.

### **Additional Notes**
- This enhancement is backward compatible and does not interfere with the existing Helm integration in Kustomize.
- The `--helm-debug` flag is optional and can be omitted if detailed Helm debugging is not required.
